lib/active_storage/service/s3_service.rb

```diff
@@ -4,84 +4,72 @@
 class ActiveStorage::Service::S3Service < ActiveStorage::Service
   attr_reader :client, :bucket
 
-  def initialize(access_key_id:, secret_access_key:, region:, bucket:)
-    @client = Aws::S3::Resource.new(access_key_id: access_key_id, secret_access_key: secret_access_key, region: region)
-    @bucket = @client.bucket(bucket)
+  def initialize(bucket:, client: nil, **client_options)
+    @bucket = bucket
+    @client = client || Aws::S3::Client.new(client_options)
   end
 
   def upload(key, io, checksum: nil)
     instrument :upload, key, checksum: checksum do
       begin
-        object_for(key).put(body: io, content_md5: checksum)
+        client.put_object bucket: bucket, key: key, body: io, content_md5: checksum
       rescue Aws::S3::Errors::BadDigest
         raise ActiveStorage::IntegrityError
       end
     end
   end
 
-  def download(key)
+  def download(key, &block)
     if block_given?
       instrument :streaming_download, key do
-        stream(key, &block)
+        client.get_object bucket: bucket, key: key, &block
       end
     else
       instrument :download, key do
-        object_for(key).get.body.read.force_encoding(Encoding::BINARY)
+        "".b.tap do |data|
+          client.get_object bucket: bucket, key: key, response_target: data
+        end
       end
     end
   end
 
   def delete(key)
     instrument :delete, key do
-      object_for(key).delete
+      client.delete_object bucket: bucket, key: key
     end
   end
 
   def exist?(key)
     instrument :exist, key do |payload|
-      answer = object_for(key).exists?
-      payload[:exist] = answer
-      answer
+      payload[:exist] =
+        begin
+          client.head_object bucket: bucket, key: key
+        rescue Aws::S3::Errors::NoSuckKey
+          false
+        else
+          true
+        end
```

Comment:
`Object#exists?` polls until the object exists or it hits a timeout, which is very slow for objects that don't actually exist. S3 has read-after-write consistency, so we don't need this in the hot path. (Eventual consistency for read-after-overwrite/delete.) Just check whether the object exists immediately instead.